src/parser/ir/ir_parser.h

```diff
@@ -37,10 +38,22 @@ using ir_parser_generator::IrParser;
 
 class IrProgramParser {
  public:
-  // Parses an input string
-  const Module& ParseProgram(absl::string_view prog_text);
+  struct Result {
+    std::unique_ptr<IRContext> context;
+    std::unique_ptr<Module> module;
+    std::unique_ptr<SsaNames> ssa_names;
+  };
+
+  static Result ParseProgram(absl::string_view prog_text) {
+    IrProgramParser parser(prog_text);
+    return {.context = std::move(parser.context_),
+            .module = std::move(parser.module_),
+            .ssa_names = std::move(parser.ssa_names_)};
+  }
```

Comment:
You will still need std::move even if you don't use unique_ptr?

In any case, we needed to return `unique_ptr` for pointer stability. Note that `ssa_names_` is keeping a map from `const Module*` to `std::string`. When we do `std::move(module)` and `std::move(ssa_names_)`, the entry for `module` in the `ssa_names_` becomes invalid. As a quick hack, I suggested using `std::unique_ptr`.

Reply:
I meant that comment in the context of my suggestion to eliminate the mutable result fields. As you pointed out, that's less straightforward than I initially thought, so this comment makes less sense than it once did.

I still think it might make sense to have one `std::unique_ptr` to a `Result` rather than three `std::unique_ptr`s to these different objects though, depending on how these are used later.

Reply:
Re:Take the work out of the constructor for IrProgramParser and put it here, making this method non-static.
@bgogul : We will be losing the static nature of the function. I remember you mentioning it would be good to have this static. Do we all agree on this?